File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/comum/rotinasdecalculo/CalculoDoIntegralizar.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.comum.rotinasdecalculo;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Periodo;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.rotinasdecalculo.RotinaDeCalculo;
import java.math.BigDecimal;

public class CalculoDoIntegralizar
extends RotinaDeCalculo {
    private static final int VINTE_OITO_DIAS = 28;
    private static final int VINTE_NOVE_DIAS = 29;
    private static final int TRINTA_DIAS = 30;
    private static final int TRINTA_UM_DIAS = 31;
    private static final String PERIODO = "PERIODO";
    private static final String VALOR = "VALOR";
    private static final String RESULTADO = "RESULTADO";
    private static final String QTD_EXCLUSOES = "QTD_EXCLUSOES";

    public CalculoDoIntegralizar(Periodo periodo, BigDecimal valor) {
        this.setEntrada(PERIODO, periodo);
        this.setEntrada(VALOR, valor);
        this.setEntrada(QTD_EXCLUSOES, 0);
    }

    public CalculoDoIntegralizar(Periodo periodo, BigDecimal valor, Integer qtdExclusoes) {
        this.setEntrada(PERIODO, periodo);
        this.setEntrada(VALOR, valor);
        this.setEntrada(QTD_EXCLUSOES, qtdExclusoes);
    }

    @Override
    public void executar() {
        BigDecimal valor = this.getEntrada(VALOR, BigDecimal.class);
        Periodo periodo = this.getEntrada(PERIODO, Periodo.class);
        int diasDoPeriodo = periodo.totalDeDias();
        int qtdExclusoes = this.getEntrada(QTD_EXCLUSOES, Integer.class);
        if ((diasDoPeriodo -= qtdExclusoes) < 0) {
            diasDoPeriodo = 0;
        }
        if (diasDoPeriodo > 30) {
            diasDoPeriodo = 30;
        }
        BigDecimal resultado = BigDecimal.ZERO;
        if (Utils.naoNulos(valor, diasDoPeriodo) && diasDoPeriodo > 0) {
            switch (HelperDate.getInstance(periodo.getFinal()).daysInMonth()) {
                case 28: {
                    resultado = valor.multiply(new BigDecimal(28), Utils.CONTEXTO_MATEMATICO).divide(new BigDecimal(diasDoPeriodo), Utils.CONTEXTO_MATEMATICO);
                    break;
                }
                case 29: {
                    resultado = valor.multiply(new BigDecimal(29), Utils.CONTEXTO_MATEMATICO).divide(new BigDecimal(diasDoPeriodo), Utils.CONTEXTO_MATEMATICO);
                    break;
                }
                default: {
                    resultado = valor.multiply(new BigDecimal(30), Utils.CONTEXTO_MATEMATICO).divide(new BigDecimal(diasDoPeriodo), Utils.CONTEXTO_MATEMATICO);
                }
            }
        }
        this.setSaida(RESULTADO, resultado);
    }

    public BigDecimal getResultado() {
        return this.getSaida(RESULTADO, BigDecimal.class);
    }
}

File: src/lib/pjecalc/core/comum/rotinasdecalculo/calculo-do-integralizar.ts
/**
 * PJe-Calc v2.15.1 — CalculoDoIntegralizar
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.comum.rotinasdecalculo.CalculoDoIntegralizar
 *
 * Ref Java: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/comum/rotinasdecalculo/CalculoDoIntegralizar.java
 *
 * Extrapola um valor proporcional (de um período parcial) para o valor "integral"
 * (como se fosse o mês completo). Usado para devidoIntegral, pagoIntegral, quantidadeIntegral.
 *
 * Fórmula: resultado = valor × diasNoMes / diasDoPeriodo
 *   - diasNoMes: 28, 29 ou 30 (nunca 31 — PJe-Calc usa 30 para meses com 30 ou 31 dias)
 *   - diasDoPeriodo: min(totalDeDias - exclusoes, 30)
 */
import Decimal from 'decimal.js';
import { HelperDate } from '../../base/comum/helper-date';
import { Periodo } from '../../base/comum/periodo';
import { naoNulos } from '../../base/comum/utils';

export class CalculoDoIntegralizar {
  private resultado: Decimal = new Decimal(0);

  constructor(
    private periodo: Periodo,
    private valor: Decimal,
    private qtdExclusoes: number = 0
  ) {}

  executar(): void {
    let diasDoPeriodo = this.periodo.totalDeDias() - this.qtdExclusoes;
    if (diasDoPeriodo < 0) diasDoPeriodo = 0;
    if (diasDoPeriodo > 30) diasDoPeriodo = 30;

    if (diasDoPeriodo > 0) {
      const diasNoMesFinal = HelperDate.getInstance(this.periodo.getFinal())!.daysInMonth();
      const diasRef = diasNoMesFinal === 28 ? 28 : diasNoMesFinal === 29 ? 29 : 30;
      this.resultado = this.valor.times(diasRef).div(diasDoPeriodo);
    } else {
      this.resultado = new Decimal(0);
    }
  }

  getResultado(): Decimal { return this.resultado; }
}
